HeartbeatLogging/Sources/HeartbeatStorage.swift

```diff
@@ -14,57 +14,86 @@
 
 import Foundation
 
-/// A type that can store and remove hearbeats.
 protocol HeartbeatStorageProtocol {
-  func offer(_ heartbeat: Heartbeat)
-  func flush() -> HeartbeatInfo?
+  typealias HeartbeatInfoTransform = (HeartbeatInfo?) -> HeartbeatInfo?
+
+  func async(_ transform: @escaping HeartbeatInfoTransform)
+  // TODO: Evaluate if async variant of below API is needed.
+  func getAndReset(using transform: HeartbeatInfoTransform?) throws -> HeartbeatInfo?
 }
 
 /// Thread-safe storage object designed for storing heartbeat data.
 final class HeartbeatStorage: HeartbeatStorageProtocol {
-  private let storage: PersistentStorage
-  // TODO: Add documentation.
+  // The identifier used to differentiate instances.
+  private let id: String
+  // The underlying storage container to read from and write to.
+  private let storage: Storage
+  // An object used to encode and decode Codable heartbeat data.
   private let coder: Coder
+  // The queue for synchronizing storage operations.
   private let queue: DispatchQueue
 
-  private let limit: Int = 25 // TODO: Decide how this will be injected...
-
-  init(id: String, // TODO: - Sanitize!
-       storage: PersistentStorage,
+  init(id: String,
+       storage: Storage,
        coder: Coder = JSONCoder(),
        queue: DispatchQueue? = nil) {
+    self.id = id
     self.storage = storage
     self.coder = coder
     self.queue = queue ?? DispatchQueue(label: "com.heartbeat.storage.\(id)")
   }
 
-  func offer(_ heartbeat: Heartbeat) {
+  // MARK: - Instance Management
+
+  /// <#Description#>
+  private(set) static var cachedInstances: [String: Weak<HeartbeatStorage>] = [:]
+
+  /// <#Description#>
+  /// - Parameter id: <#id description#>
+  /// - Returns: <#description#>
+  static func getInstance(id: String) -> HeartbeatStorage {
```

Comment:
FYI. We had a discussion with @VinayGuthal and @vkryachko yesterday and they mostly convinced me that a case when we have two Firebase app instances with the same app ID is very rare, so we can ignore it. If we don't account for this case then we can have an instance of storage Firebase app instance which may simplify the implementation (it was the case for Android). In this case storage would need to be keyed by the app ID and app name. No immediate implications for Apple SDK so far, but if implementation becomes too complex, feel free to apply this simplification as well.